File: scripts/Controller.py
#!/Users/subhodeep/venv/bin/python3
'''classes for MPC'''
import numpy as np
from dataclasses import dataclass
from typing import Tuple, Optional
from CartesianFrenetConverter import CartesianFrenetConverter
try:
    from scipy.optimize import minimize
except ImportError as e:
    raise ImportError("This controller requires SciPy for optimization: pip install scipy") from e

@dataclass
class MPCParams:
    dt: float = 0.1                  # sampling time [s]
    N: int = 15                      # prediction horizon
    v_min: float = -0.8              # min linear velocity [m/s]
    v_max: float = 1.0               # max linear velocity [m/s]
    w_min: float = -2.5              # min angular velocity [rad/s]
    w_max: float = 2.5               # max angular velocity [rad/s]
    dv_max: float = 0.5              # max |Δv| per step [m/s]
    dw_max: float = 1.0              # max |Δω| per step [rad/s]
    # costs
    q_xy: float = 2.0                # weight on (x,y) position error
    q_th: float = 0.3                # weight on heading error
    r_v: float = 0.02                # weight on absolute v
    r_w: float = 0.02                # weight on absolute ω
    s_v: float = 0.5                 # weight on |Δv|
    s_w: float = 0.3                 # weight on |Δω|
    qT_xy: float = 6.0               # terminal (x,y)
    qT_th: float = 1.0               # terminal heading


class UnicycleMPC:
    """
    Model Predictive Controller for a unicycle robot.
    State:   x = [x, y, theta]
    Control: u = [v, w]  (linear, angular)
    Dynamics:
        x_{k+1} = x_k + dt * [ v*cos(theta_k), v*sin(theta_k), w ]
    """
    def __init__(self, params: Optional[MPCParams] = None):
        self.p = params or MPCParams()
        # Keep last solution for warm-start
        self._u_prev_seq = np.zeros(2 * self.p.N)  # [v0..vN-1, w0..wN-1]

    # ----------------- public API -----------------
    def control(self, x_now: np.ndarray, x_target: np.ndarray) -> Tuple[float, float]:
        """
        Compute the first control action (v, w).
        x_now:     [x, y, theta]
        x_target:  [x*, y*, theta*]  (theta* optional: if NaN, ignore heading error)
        """
        x_now = np.asarray(x_now, dtype=float).reshape(3)
        x_ref = np.asarray(x_target, dtype=float).reshape(3)
        theta_ref_valid = np.isfinite(x_ref[2])

        # Warm start: shift previous optimal controls forward one step
        u0 = self._warm_start()

        # Box bounds on controls
        bounds = [(self.p.v_min, self.p.v_max)] * self.p.N + [(self.p.w_min, self.p.w_max)] * self.p.N

        # Build inequality constraints for slew-rate (Δu) limits: |u_k - u_{k-1}| <= d_max
        A_ineq, b_ineq = self._delta_constraints()

        cons = []
        if A_ineq is not None:
            cons.append({
                "type": "ineq",
                "fun": lambda u, A=A_ineq, b=b_ineq: b - A @ u
            })

        # Objective
        def objective(u: np.ndarray) -> float:
            v_seq = u[:self.p.N]
            w_seq = u[self.p.N:]
            # simulate
            traj = self._rollout(x_now, v_seq, w_seq)
            # stage costs
            pos_err = traj[:, :2] - x_ref[:2]          # (N+1, 2)
            e_xy2 = np.sum(pos_err**2, axis=1)        # (N+1,)
            if theta_ref_valid:
                e_th = np.array([CartesianFrenetConverter.normalize_angle(th - x_ref[2]) for th in traj[:, 2]])
                e_th2 = e_th**2
            else:
                e_th2 = np.zeros(self.p.N + 1)

            # smoothness (Δu) and effort
            dv = np.diff(v_seq, prepend=v_seq[0])
            dw = np.diff(w_seq, prepend=w_seq[0])

            J_stage = np.sum(self.p.q_xy * e_xy2[:-1] + self.p.q_th * e_th2[:-1]
                             + self.p.r_v * (v_seq**2) + self.p.r_w * (w_seq**2)
                             + self.p.s_v * (np.abs(dv)) + self.p.s_w * (np.abs(dw)))

            # terminal
            J_term = self.p.qT_xy * e_xy2[-1] + self.p.qT_th * e_th2[-1]
            return J_stage + J_term

        res = minimize(
            objective, u0, method="SLSQP", bounds=bounds, constraints=cons,
            options={"maxiter": 100, "ftol": 1e-4, "disp": False}
        )

        u_opt = res.x if res.success else u0
        self._u_prev_seq = u_opt.copy()

        v_cmd = float(np.clip(u_opt[0], self.p.v_min, self.p.v_max))
        w_cmd = float(np.clip(u_opt[self.p.N], self.p.w_min, self.p.w_max))
        return [v_cmd, w_cmd]

    # ----------------- helpers -----------------
    def _rollout(self, x0: np.ndarray, v_seq: np.ndarray, w_seq: np.ndarray) -> np.ndarray:
        """Simulate N steps; return (N+1, 3) states including x0."""
        dt = self.p.dt
        x = x0.copy()
        traj = [x.copy()]
        for k in range(self.p.N):
            v = v_seq[k]
            w = w_seq[k]
            x[0] = x[0] + dt * v * np.cos(x[2])
            x[1] = x[1] + dt * v * np.sin(x[2])
            x[2] = CartesianFrenetConverter.normalize_angle(x[2] + dt * w)
            traj.append(x.copy())
        return np.vstack(traj)

    def _warm_start(self) -> np.ndarray:
        """Shift previous sequence forward; keep last value for the tail."""
        u = self._u_prev_seq
        N = self.p.N
        v_seq = u[:N]
        w_seq = u[N:]
        v_ws = np.r_[v_seq[1:], v_seq[-1]]
        w_ws = np.r_[w_seq[1:], w_seq[-1]]
        return np.r_[v_ws, w_ws]

    def _delta_constraints(self):
        """
        Build |Δv_k|<=dv_max, |Δw_k|<=dw_max as linear inequalities:
            -dv_max <= v_k - v_{k-1} <= dv_max   (similar for w)
        SLSQP accepts A u <= b. We express both sides by stacking.
        Use first element relative to previous optimal first element (fixed at itself -> no constraint tightening).
        """
        N = self.p.N
        dv = self.p.dv_max
        dw = self.p.dw_max
        if not np.isfinite(dv) and not np.isfinite(dw):
            return None, None

        # Variables: u = [v0..vN-1, w0..wN-1]
        dim = 2 * N
        rows = []
        rhs = []

        # v deltas
        if np.isfinite(dv):
            # k=0: we don't constrain against unknown previous command; instead cap |v0| change by bounds already.
            for k in range(1, N):
                row = np.zeros(dim)
                row[k] = 1.0      # v_k
                row[k - 1] = -1.0 # -v_{k-1}
                rows.append(row.copy());  rhs.append(dv)   #  v_k - v_{k-1} <= dv
                rows.append(-row.copy()); rhs.append(dv)   # -v_k + v_{k-1} <= dv  => -(v_k - v_{k-1}) <= dv

        # w deltas
        if np.isfinite(dw):
            offset = N
            for k in range(1, N):
                row = np.zeros(dim)
                row[offset + k] = 1.0
                row[offset + k - 1] = -1.0
                rows.append(row.copy());  rhs.append(dw)
                rows.append(-row.copy()); rhs.append(dw)

        if not rows:
            return None, None
        A = np.vstack(rows)
        b = np.asarray(rhs)
        return A, b

File: src/Controller.cpp
/*
    Script for controller class
    Author: Subhodeep Choudhury
    Source: https://github.com/SnyprDragun/Motion-planning-Frenet-frame
*/

#include "Controller.hpp"

Controller::Controller(const MPCParams& params) : p(params) {
    u_prev_seq = VectorXd::Zero(2 * p.N);
}

vector<float> Controller::control(const PathPoint& x_now, const PathPoint& x_target) {
    VectorXd x(3);
    x << x_now.x, x_now.y, x_now.theta;
    VectorXd x_ref(3);
    x_ref << x_target.x, x_target.y, x_target.theta;

    // Warm start
    VectorXd u0 = warm_start();

    // NLopt setup
    nlopt::opt opt(nlopt::LD_SLSQP, 2 * p.N);
    vector<double> lb(2*p.N), ub(2*p.N);
    for (int i=0;i<p.N;i++) { lb[i]=p.v_min; ub[i]=p.v_max; }
    for (int i=p.N;i<2*p.N;i++) { lb[i]=p.w_min; ub[i]=p.w_max; }
    opt.set_lower_bounds(lb);
    opt.set_upper_bounds(ub);

    ObjData objdata{this, x, x_ref};
    opt.set_min_objective(objective_wrapper, &objdata);
    opt.set_maxeval(100);
    opt.set_ftol_rel(1e-4);

    vector<double> u_vec(u0.data(), u0.data()+u0.size());
    double minf;
    nlopt::result result = nlopt::FAILURE;
    try {
        result = opt.optimize(u_vec, minf);
    } catch(...) {
        u_vec.assign(u0.data(), u0.data()+u0.size()); // fallback
    }

    u_prev_seq = Map<VectorXd>(u_vec.data(), u_vec.size());

    auto clamp = [](float val, float min_val, float max_val) {
        return max(min_val, min(val, max_val));
    };
    float v_cmd = clamp(u_prev_seq(0), p.v_min, p.v_max);
    float w_cmd = clamp(u_prev_seq(p.N), p.w_min, p.w_max);
    cout << "Control commands: v = " << v_cmd << ", w = " << w_cmd << endl;
    return {v_cmd, w_cmd};
}

MatrixXd Controller::rollout(const VectorXd& x0, const VectorXd& v_seq, const VectorXd& w_seq) {
    MatrixXd traj(p.N+1, 3);
    VectorXd x = x0;
    traj.row(0) = x.transpose();
    for (int k=0;k<p.N;k++) {
        double v = v_seq(k);
        double w = w_seq(k);
        x(0) += p.dt * v * cos(x(2));
        x(1) += p.dt * v * sin(x(2));
        x(2) = normalize_angle(x(2) + p.dt*w);
        traj.row(k+1) = x.transpose();
    }
    return traj;
}

VectorXd Controller::warm_start() {
    VectorXd v_seq = u_prev_seq.head(p.N);
    VectorXd w_seq = u_prev_seq.tail(p.N);
    VectorXd v_ws(p.N), w_ws(p.N);
    v_ws.head(p.N-1) = v_seq.tail(p.N-1);
    v_ws(p.N-1) = v_seq(p.N-1);
    w_ws.head(p.N-1) = w_seq.tail(p.N-1);
    w_ws(p.N-1) = w_seq(p.N-1);
    VectorXd u_ws(2*p.N);
    u_ws << v_ws, w_ws;
    return u_ws;
}

double Controller::normalize_angle(double th) {
    while(th > M_PI) th -= 2*M_PI;
    while(th < -M_PI) th += 2*M_PI;
    return th;
}

double Controller::objective_wrapper(const vector<double>& u, vector<double>& grad, void* data) {
    ObjData* d = static_cast<ObjData*>(data);
    VectorXd u_vec = Map<const VectorXd>(u.data(), u.size());
    VectorXd v_seq = u_vec.head(d->mpc->p.N);
    VectorXd w_seq = u_vec.tail(d->mpc->p.N);

    MatrixXd traj = d->mpc->rollout(d->x_now, v_seq, w_seq);

    // Stage costs
    VectorXd pos_err2(traj.rows());
    for (int k=0;k<traj.rows();k++)
        pos_err2(k) = pow(traj(k,0)-d->x_target(0),2) + pow(traj(k,1)-d->x_target(1),2);

    VectorXd e_th2(traj.rows());
    for (int k=0;k<traj.rows();k++)
        e_th2(k) = pow(d->mpc->normalize_angle(traj(k,2)-d->x_target(2)),2);

    VectorXd dv = v_seq;
    dv.tail(dv.size()-1) -= v_seq.head(dv.size()-1);
    dv(0) = 0.0;
    VectorXd dw = w_seq;
    dw.tail(dw.size()-1) -= w_seq.head(dw.size()-1);
    dw(0) = 0.0;

    double J_stage = 0.0;
    for (int k=0;k<d->mpc->p.N;k++) {
        J_stage += d->mpc->p.q_xy * pos_err2(k) + d->mpc->p.q_th * e_th2(k)
                 + d->mpc->p.r_v * pow(v_seq(k),2) + d->mpc->p.r_w * pow(w_seq(k),2)
                 + d->mpc->p.s_v * abs(dv(k)) + d->mpc->p.s_w * abs(dw(k));
    }

    double J_term = d->mpc->p.qT_xy * pos_err2(d->mpc->p.N) + d->mpc->p.qT_th * e_th2(d->mpc->p.N);
    return J_stage + J_term;
}
